File: projects/06/ASSEMBLER/src/Parser.java
import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;

public class Parser {
	
	public static final int A_COMMAND = 0;
	public static final int C_COMMAND = 1;
	public static final int L_COMMAND = 2;
	private static final int CANTLIMITECARACTERES = 1000000;
	public String lineaActual;
	public int nroDeLineaActual = 0;

	private BufferedReader archivoAsm;
	

	Parser(String archivo){

		try{
			FileReader fileReader = new FileReader(archivo);
			archivoAsm = new BufferedReader(fileReader); 
			lineaActual = archivoAsm.readLine();
            lineaActual = lineaActual.trim();
			archivoAsm.mark(CANTLIMITECARACTERES);
		}
		catch (FileNotFoundException e){
			System.out.println("Error. No se encontro el archivo " + archivo);
		} catch (IOException e) {
			System.out.println("Error. No se pudo leer el archivo" + archivo);
		}
		
	}

	public boolean hasMoreCommands() throws IOException{
		archivoAsm.mark(CANTLIMITECARACTERES);
		String lineaSiguiente = archivoAsm.readLine();

		if (lineaSiguiente == null){
			archivoAsm.reset();
			return false;
		}
		return true;
	}

	/**
	 * Avanza hasta el siguiente comando de assembly, ignorando las lineas
	 * vacias y lineas de comentarios
	 * @throws IOException
	 */
	public void advance() throws IOException{
		lineaActual = archivoAsm.readLine();
		if (lineaActual == null){
			return;
		}
		lineaActual= lineaActual.trim();
		if ( lineaActual.isEmpty() || lineaActual.startsWith("//") ){
			this.advance();
		} else {
			nroDeLineaActual++;
		}

	}
	
	public int commandType(){
		return 0;
		
	}
	
	public String symbol(){
		return lineaActual;
		
	}

	public String dest(){
		return lineaActual;
		
	}

	public String comp(){
		return lineaActual;
		
	}

	public String jump(){
		return lineaActual;
		
	}
	
}

File: projects/06/assembler/parser.py
class Parser():

    def __init__(self, file):
        content = [line.strip() for line in file.readlines()]
        content = [line.split('//')[0].strip() for line in content if not line.startswith('//') and line is not '']
        self.source_code: list = content
        self.current_command: int = 0

    def _command(self):
        return self.source_code[self.current_command]

    def has_more_commands(self):
        if self.current_command < len(self.source_code):
            return True
        return False

    def advance(self):
        if self.has_more_commands():
            self.current_command += 1

    def command_type(self):
        if self._command().startswith('@'):
            return 'A_COMMAND'
        if self._command().startswith('('):
            return 'L_COMMAND'
        return 'C_COMMAND'

    def symbol(self):
        command = self._command()
        if command.startswith('@'):
            return command.split('@')[1]
        return command[command.find('(') + 1: command.find(')')]

    def dest(self):
        command = self._command()
        if command.find('=') is -1:
            return ''
        return command[0:command.find('=')]

    def comp(self):
        command = self._command()
        if command.find('=') is -1 and command.find(';') is -1:
            return command
        if command.find('=') is -1 and command.find(';') is not -1:
            return command[:command.find(';')]
        if command.find('=') is not -1 and command.find(';') is -1:
            return command[command.find('=') + 1:]
        return command[command.find('=') + 1:command.find(';')]

    def jump(self):
        command = self._command()
        if command.find(';') is -1:
            return ''
        return command[command.find(';') + 1:]
